### PR TITLE
CLI events

### PR DESCRIPTION
**Pull request by dnil**
This PR adds a functionality or fixes a bug.

Screenshot 2024-05-16 at 15 43 27

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. load a case from the CLI
2. note a CLI user event in the case audit log
3. update individual and have it work as before
4. scout update case from the CLI, e.g. adding a collaborator, and note a corresponding event in the case audit log

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

File: scout/server/blueprints/cases/controllers.py [hunk 1]
```diff
@@ -805,27 +805,6 @@ def update_synopsis(store, institute_obj, case_obj, user_obj, new_synopsis):
         store.update_synopsis(institute_obj, case_obj, user_obj, link, content=new_synopsis)
 
 
-def _update_case(store, case_obj, user_obj, institute_obj, verb):
```
Comment (dnil) [inline scout/server/blueprints/cases/controllers.py:808]: Moved to CaseEventHandler. Does create the events already. 🙄

File: scout/server/blueprints/cases/controllers.py [hunk 2]
```diff
@@ -841,8 +820,13 @@ def update_individuals(store, institute_obj, case_obj, user_obj, ind, age, tissu
 
     case_obj["individuals"] = case_individuals
 
-    verb = "update_individual"
-    _update_case(store, case_obj, user_obj, institute_obj, verb)
```
Comment (dnil) [inline scout/server/blueprints/cases/controllers.py:845]: Changed to not pass the verb, but instead use two different EventHandler functions. Also create link externally, like for most events.

File: tests/adapter/mongo/test_sanger_validation.py [hunk 3]
```diff
@@ -99,6 +99,7 @@ def test_get_sanger_unevaluated(
     ## Prepare for variants for Sanger assignment
     # Check that variant collections is empty
     assert sum(1 for i in adapter.variant_collection.find()) == 0
+    previous_events = sum(1 for i in adapter.event_collection.find())
```
Comment (dnil) [inline tests/adapter/mongo/test_sanger_validation.py:102]: Changed several tests that rely on event_collection to be empty after setup.

File: tests/load/test_load_case.py
Comment (dnil) [inline tests/load/test_load_case.py:1]: I found no uses of this. Perhaps was part of some very old interface imported into a monolith CG.

**Comment by northwestwitch** [comment]
It's one of those..

image

**Comment by dnil** [comment]
> It's one of those..
> 
> image

I figured it looked so empty on the PR list.. 😁

**Comment by dnil** [comment]
I'm setting it as ready for review, but dont feel you must spend the Friday afternoon on it; this has been sitting around on issues for a long time. It isn't that huge a thing though, once you look at it: it is mostly a couple of helper functions that have moved from the CaseHandler to the CaseEventHandler, now creating events for the user to see, and the calls for them updated accordingly. However, since we now get an event when creating a case using the load or update, a lot of tests needed updating. And when touching those functions, a few en passant things were highlighted.

I think this can be pretty useful both for proper audit and hopefully lowers the mental churn on debugging weird stuff involving case load and update.

File: scout/commands/update/case.py [hunk 4]
```diff
@@ -94,7 +94,6 @@ def case(
     """
     Update a case in the database
     """
-    adapter = store
```
Comment (dnil) [inline scout/commands/update/case.py:97]: Search and replace has been invented.

File: scout/adapter/mongo/case_events.py [hunk 5]
```diff
@@ -1,17 +1,103 @@
...
 
 class CaseEventHandler(object):
     """Class to handle case events for the mongo adapter"""
 
+    def get_cli_user(self) -> dict:
+        """
+        Return a faux CLI user with a login username from OS CLI if it is available.
+        """
+        try:
+            cli_user_name = getlogin()
```
Comment (northwestwitch) [inline scout/adapter/mongo/case_events.py:22]: Cool!

image

File: scout/adapter/mongo/case_events.py [hunk 6]
```diff
@@ -1,17 +1,86 @@
...
+            user=self.get_cli_user(),
+            link=link,
+            category="case",
+            verb="add_case",
+            subject=case_obj["display_name"],
+        )
+
+        return self.case_collection.insert_one(case_obj)
+
+    def update_case_individual(
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/adapter/mongo/case_events.py:51]: 🥇

File: scout/commands/update/case.py (hunk 4 above)
Comment (northwestwitch) [reply scout/commands/update/case.py:97]: 💯

File: scout/server/blueprints/variants/controllers.py [hunk 7]
```diff
@@ -1996,7 +1996,7 @@ def activate_case(store, institute_obj, case_obj, current_user):
         store.update_status(institute_obj, case_obj, user_obj, "active", case_link)
 
 
-def reset_all_dimissed(store, institute_obj, case_obj):
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1999]: 😆

File: tests/adapter/mongo/test_case_events.py [hunk 8]
```diff
@@ -18,7 +18,7 @@ def test_case_dismissed_variants(adapter, institute_obj, case_obj, user_obj, var
         dismiss_variant=dismiss_reason,
     )
     # THEN one event should be found in events collection
-    assert sum(1 for i in adapter.event_collection.find()) == 1
+    assert sum(1 for _ in adapter.event_collection.find()) == 1
```
Comment (northwestwitch) [inline tests/adapter/mongo/test_case_events.py:21]: 👍🏻

File: tests/adapter/mongo/test_case_handling.py [hunk 9]
```diff
@@ -25,15 +25,44 @@ def test_add_cases(adapter, case_obj):
...
+    adapter.add_case(case_obj, institute_obj)
+
+    ## THEN assert that the case have been loaded with rank_model
+    loaded_case = adapter.case_collection.find_one({"_id": case_obj["_id"]})
+
+    assert loaded_case["rank_model_version"] == case_obj["rank_model_version"]
+    assert loaded_case["sv_rank_model_version"] == case_obj["sv_rank_model_version"]
+
+
+def test_add_case_limsid(case_obj, institute_obj, adapter):
```
Comment (northwestwitch) [inline tests/adapter/mongo/test_case_handling.py:53]:
```suggestion
def test_add_case_lims_id(case_obj, institute_obj, adapter):
```

Comment (northwestwitch) [reply tests/adapter/mongo/test_case_handling.py:53]: 🥇

**Review by northwestwitch — APPROVED** [review]
Great added feature! 
You were right, most of the changes were _ instead of i.

Very nice also with the more accurate tests

**Pull request merged**